dhis-2/dhis-services/dhis-service-expression-parser/src/main/java/org/hisp/dhis/expressionparser/DefaultExpressionParserService.java

```diff
@@ -0,0 +1,322 @@
+
+    private static Cache<String, ParseTree> EXPRESSION_PARSE_TREES = Caffeine.newBuilder()
+        .expireAfterAccess( 10, TimeUnit.MINUTES ).initialCapacity( 10000 )
+        .maximumSize( 50000 ).build();
+
+    // -------------------------------------------------------------------------
+    // Expression methods
+    // -------------------------------------------------------------------------
+
+    @Override
+    public Set<DimensionalItemObject> getExpressionDimensionalItemObjects( String expression )
+    {
+        if ( expression == null )
+        {
+            return new HashSet<>();
+        }
+
+        ExpressionItemsVisitor expressionItemsVisitor = new ExpressionItemsVisitor();
+
+        ParseTree parseTree = getParseTree( expression, true );
+
+        if ( parseTree != null )
+        {
+            try
+            {
+                return expressionItemsVisitor.getDimensionalItemObjects( parseTree,
+                    dimensionService );
+            }
+            catch ( ExpressionParserException ex )
+            {
+                log.warn( ex.getMessage() + " getting value of expression '" + expression + "'" );
+            }
+        }
+
+        return new HashSet<>();
+    }
+
+    @Override
+    public Set<OrganisationUnitGroup> getExpressionOrgUnitGroups( String expression )
+    {
+        if ( expression == null )
+        {
+            return new HashSet<>();
+        }
+
+        ExpressionItemsVisitor expressionItemsVisitor = new ExpressionItemsVisitor();
+
+        ParseTree parseTree = getParseTree( expression, true );
+
+        if ( parseTree != null )
+        {
+            try
+            {
+                return expressionItemsVisitor.getOrgUnitGroups(
+                    parseTree, organisationUnitGroupService );
+            }
+            catch ( ExpressionParserException ex )
+            {
+                log.warn( ex.getMessage() + " getting organisation unit groups of expression '" + expression + "'" );
+            }
+        }
+
+        return new HashSet<>();
+    }
+
+    @Override
+    public Double getExpressionValue( String expression,
+        Map<DimensionalItemObject, Double> valueMap, Map<String, Double> constantMap,
+        Map<String, Integer> orgUnitCountMap, Integer days )
+    {
+        if ( expression == null )
+        {
+            return null;
+        }
+
+        ExpressionValueVisitor expressionValueVisitor = new ExpressionValueVisitor();
+
+        ParseTree parseTree = getParseTree( expression, true );
+
+        if ( parseTree == null )
+        {
+            return null;
+        }
+
+        try
+        {
+            return expressionValueVisitor.getExpressionValue( parseTree, valueMap,
+                    constantMap, orgUnitCountMap, days );
+        }
+        catch ( ExpressionParserException ex )
+        {
+            log.warn( ex.getMessage() + " finding items in expression '" + expression + "'" );
```

Comment:
Should we use the log.warn( String, Throwable ) method to make the exception more traceable? There might be a good reason for not doing that here.

Reply:
I'm not sure. I can add that if you want. I had been thinking that user errors (like importing bad expressions) should not result in stack traces, whereas unexpected code execution should. Do we have a guiding principle for this?